lib/puppet/util/windows/user.rb

```diff
@@ -145,6 +145,90 @@ def load_profile(user, password)
+
+    result = LsaEnumerateAccountRights(policy_handle.read_pointer, sid_pointer, rights_pointer, number_of_rights)
+
+    win32_error = LsaNtStatusToWinError(result)
+    return rights if win32_error == 2 # No rights found error
+
+    check_lsa_nt_status_and_raise_failures(result, "`LsaEnumerateAccountRights` method failed")
+
+    # https://stackoverflow.com/questions/24461466/how-to-use-ruby-ffi-to-read-array-of-static-structures
+    number_of_rights.read_ulong.times do |i|
+      right = LSA_UNICODE_STRING.new(rights_pointer.read_pointer + i * LSA_UNICODE_STRING.size)
+      rights << right[:Buffer].read_arbitrary_wide_string_up_to
+    end
+
+    result = LsaFreeMemory(rights_pointer.read_pointer)
+    check_lsa_nt_status_and_raise_failures(result, "`LsaFreeMemory` method failed")
+
+    LsaClose(policy_handle.read_pointer)
+
+    rights
+  end
+  module_function :get_rights
+
+  def set_right(name, right)
+    right_pointer = new_right_pointer(right)
+    policy_handle = open_new_lsa_policy
+    sid_pointer = Puppet::Util::Windows::SID::Principal::lookup_sid_pointer(name)
+
+    result = LsaAddAccountRights(policy_handle.read_pointer, sid_pointer, right_pointer, 1)
+    check_lsa_nt_status_and_raise_failures(result, "`LsaAddAccountRights` method failed")
+
+    LsaClose(policy_handle.read_pointer)
+  end
+  module_function :set_right
+
+  def remove_right(name, right)
+    right_pointer = new_right_pointer(right)
+    policy_handle = open_new_lsa_policy
+    sid_pointer = Puppet::Util::Windows::SID::Principal::lookup_sid_pointer(name)
+
+    result = LsaRemoveAccountRights(policy_handle.read_pointer, sid_pointer, false, right_pointer, 1)
+    check_lsa_nt_status_and_raise_failures(result, "`LsaRemoveAccountRights` method failed")
+
+    LsaClose(policy_handle.read_pointer)
+  end
+  module_function :remove_right
+
+  def self.open_new_lsa_policy
+    access = 0
+    access |= POLICY_CREATE_ACCOUNT
+    access |= POLICY_LOOKUP_NAMES
+
+    policy_handle = FFI::MemoryPointer.new(:pointer)
+    result = LsaOpenPolicy(nil, LSA_OBJECT_ATTRIBUTES.new, access, policy_handle)
+    check_lsa_nt_status_and_raise_failures(result, "`LsaOpenPolicy` method failed")
+
+    policy_handle
+  end
+  private_class_method :open_new_lsa_policy
+
+  def self.new_right_pointer(right)
+    right_pointer = FFI::MemoryPointer.new LSA_UNICODE_STRING, 1
+    right_lsa_string = LSA_UNICODE_STRING.new(right_pointer)
+    right_lsa_string[:Buffer] = FFI::MemoryPointer.from_string(wide_string(right))
+    right_lsa_string[:Length] = right.length * 2
+    right_lsa_string[:MaximumLength] = right.length * 2 + 2
+
+    right_pointer
+  end
+  private_class_method :new_right_pointer
+
+  def self.check_lsa_nt_status_and_raise_failures(status, error_message)
+    win32_error = LsaNtStatusToWinError(status)
+    raise Puppet::Error.new("#{error_message}: #{win32_error}") unless win32_error == 0
```

Comment:
Might need to add some special cases. So far error codes 2 (when no rights are set) and 5 (when not running as admin) were found.